### PR TITLE
Fix #14633: Update all lockfiles in multi-module Gradle projects

### LINKED ISSUE #14633: Gradle lockfiles are not updated in multi-module projects
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Package ecosystem

Gradle

### Package manager version

9.4.1

### Language version

_No response_

### Manifest location and content before the Dependabot update

/app/build.gradle
/app/gradle.lockfile

### dependabot.yml content

/.github/dependabot.yml

### Updated dependency

_No response_

### What you expected to see, versus what you actually saw

I expect Dependabot to update the Gradle dependency _and_ the lockfile, but it only updated the dependency.

### Native package manager behavior

_No response_

### Images of the diff or a link to the PR, issue, or logs

_No response_

### Smallest manifest that reproduces the issue

https://github.com/nikammerlaan/gradle-lockfile-dependabot-repro

### PR DESCRIPTION
**Pull request by v-HaripriyaC**
## What are you trying to accomplish?

Fixes #14633 — Gradle lockfiles in submodules are not regenerated when dependencies change in multi-module projects. 

Previously, the lockfile updater only selected lockfiles in the same directory as the buildfile being processed. This meant when a parent module's dependencies changed, lockfiles in sibling submodules would not be updated.

### Before:
- `gradle dependencies --write-locks` runs with `file.directory == build_file.directory` filter
- Only lockfiles in same directory as current buildfile are selected
- Submodule lockfiles remain stale

### After:
- Select ALL lockfiles regardless of directory
- Run gradle from project root via `settings.gradle` detection
- Gradle automatically updates all lockfiles in all modules

## Anything you want to highlight for special attention from reviewers?

• **Simplified approach**: Instead of complex subproject task prefixing, we leverage Gradle's built-in multi-module support. Running `gradle dependencies --write-locks` from the project root automatically regenerates all lockfiles.

• **Key changes**:
  - Changed lockfile selection from `file.directory == build_file.directory && file.name.end_with?(".lockfile")` to just `file.name.end_with?(".lockfile")` (line 23)
  - Detect project root by finding `settings.gradle` or `settings.gradle.kts` in dependency files (lines 33-39)
  - Added file existence check to handle projects without resolvable dependencies (line 73)
  - Added content equality check to skip unnecessary diffs (line 76)

• **No breaking changes**: Single-module projects continue to work identically; multi-module projects now work correctly.

• **Defensive programming**: Handles edge cases like version catalogs and composite builds through existing Gradle mechanisms.

## How will you know you've accomplished your goal?

• All existing unit tests pass (34 file updater tests confirmed)
• New lockfile updater tests pass (14 tests, 0 failures, 1 pending)
• Multi-module scenario tests validate the fix (app + lib + root lockfiles)
• Version catalog test confirms issue #14788 compatibility
• No regressions in other ecosystems (560 gradle tests, 6 pre-existing timezone failures unrelated to this fix)
• Syntax validation and RuboCop checks pass

## Checklist

- [x] I have run the complete test suite to ensure all tests and linters pass
- [x] I have thoroughly tested my code changes to ensure they work as expected
- [x] I have written clear and descriptive commit messages
- [x] I have provided a detailed description of the changes in the pull request
- [x] I have ensured that the code is well-documented and easy to understand
- [x] No unnecessary changes or new helper methods introduced (simplified from initial approach)

### CODE REVIEW

File: gradle/spec/dependabot/gradle/file_updater/lockfile_updater_spec.rb [hunk 1]
```diff
@@ -0,0 +1,226 @@
+# typed: false
+# frozen_string_literal: true
+
+require "spec_helper"
+require "dependabot/dependency_file"
```
[the remark this answers is not in the document]
Comment (HaripriyaChintalapati) [reply gradle/spec/dependabot/gradle/file_updater/lockfile_updater_spec.rb:5]: but those are already added on line 1 and 2

File: gradle/lib/dependabot/gradle/file_updater/lockfile_updater.rb [hunk 2]
```diff
@@ -108,6 +116,26 @@ def write_properties_file(file_name) # rubocop:disable Metrics/PerceivedComplexi
 
         sig { returns(T::Array[Dependabot::DependencyFile]) }
         attr_reader :dependency_files
+
+        sig { params(build_file: Dependabot::DependencyFile).returns(String) }
+        def determine_root_dir(build_file)
+          settings_file = find_settings_file
+          return settings_file.directory if settings_file
+
+          build_file.directory
```
[the remark this answers is not in the document]
Comment (HaripriyaChintalapati) [reply gradle/lib/dependabot/gradle/file_updater/lockfile_updater.rb:121]: This comment is not applicable. DependencyFile#directory represents the project root (job source directory), which is exactly where Gradle needs to run for multi-module lockfile regeneration. Using File.dirname(path) would run Gradle from a submodule directory — that's the bug this PR fixes (#14633). The find_settings_file method already ensures we don't pick an included build's settings file via the !f.name.include?("/") preference.

File: gradle/lib/dependabot/gradle/file_updater/lockfile_updater.rb [hunk 3]
```diff
@@ -52,9 +48,9 @@ def update_lockfiles(build_file)
 
             Dir.chdir(cwd) do
               SharedHelpers.run_shell_command(command, cwd: cwd)
-              update_lockfiles_content(temp_dir, local_lockfiles, updated_files)
+              update_lockfiles_content(temp_dir, lockfiles, updated_files)
             rescue SharedHelpers::HelperSubprocessFailed => e
-              puts "Failed to update lockfiles: #{e.message}"
+              Dependabot.logger.error("Failed to update lockfiles: #{e.message}")
```
[the remark this answers is not in the document]
Comment (HaripriyaChintalapati) [reply gradle/lib/dependabot/gradle/file_updater/lockfile_updater.rb:53]: this is not required

File: gradle/spec/dependabot/gradle/file_updater/lockfile_updater_spec.rb [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+        )
+      end
+
+      let(:version_catalog) do
+        Dependabot::DependencyFile.new(
+          name: "gradle/libs.versions.toml",
+          directory: "/",
+          content: "[versions]\njunit = \"4.13.2\"\n"
+        )
+      end
```
[the remark this answers is not in the document]
Comment (HaripriyaChintalapati) [reply gradle/spec/dependabot/gradle/file_updater/lockfile_updater_spec.rb:238]: invalid

File: gradle/spec/dependabot/gradle/file_updater/lockfile_updater_spec.rb [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+          name: "lib/build.gradle",
+          directory: "/",
+          content: fixture("buildfiles", "basic_build.gradle")
+        )
+      end
+
+      let(:lib_lockfile) do
+        Dependabot::DependencyFile.new(
+          name: "lib/gradle.lockfile",
+          directory: "/",
```
[the remark this answers is not in the document]
Comment (HaripriyaChintalapati) [reply gradle/spec/dependabot/gradle/file_updater/lockfile_updater_spec.rb:125]: invalid comment, In Dependabot, directory is the job source dir (from dependabot.yml), and name is the relative path from there. All files fetched for a single Dependabot job share the same directory. The existing Gradle specs in this repo use the same pattern.

File: gradle/spec/dependabot/gradle/file_updater/lockfile_updater_spec.rb [hunk 6]
```diff
@@ -0,0 +1,298 @@
+# typed: false
+# frozen_string_literal: true
+
+require "spec_helper"
+require "dependabot/dependency_file"
+require "dependabot/dependency"
+require "dependabot/shared_helpers"
+require "dependabot/gradle/file_updater/lockfile_updater"
```
[the remark this answers is not in the document]
Comment (HaripriyaChintalapati) [reply gradle/spec/dependabot/gradle/file_updater/lockfile_updater_spec.rb:8]: FileUtils is transitive, invalid comment

File: gradle/spec/dependabot/gradle/file_updater/lockfile_updater_spec.rb [hunk 7]
```diff
@@ -0,0 +1,298 @@
+# typed: false
+# frozen_string_literal: true
+
+require "spec_helper"
+require "dependabot/dependency_file"
+require "dependabot/dependency"
```
[the remark this answers is not in the document]
Comment (HaripriyaChintalapati) [reply gradle/spec/dependabot/gradle/file_updater/lockfile_updater_spec.rb:6]: note required / invalid

File: gradle/spec/dependabot/gradle/file_updater/lockfile_updater_spec.rb [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+      it "skips update when lockfile content is unchanged" do
+        allow(Dependabot::SharedHelpers).to receive(:run_shell_command) do
+          File.write(File.join(Dir.pwd, "gradle.lockfile"), "# lockfile content")
+        end
+
+        result = lockfile_updater.update_lockfiles(buildfile)
+
+        updated_lockfile = result.find { |f| f.name == "gradle.lockfile" }
+        # Content unchanged, so original file object is preserved
+        expect(updated_lockfile).to eq(lockfile)
```
[the remark this answers is not in the document]
Comment (HaripriyaChintalapati) [reply gradle/spec/dependabot/gradle/file_updater/lockfile_updater_spec.rb:65]: eq is correct here, not updating the code

File: gradle/lib/dependabot/gradle/file_updater/lockfile_updater.rb [hunk 9]
```diff
@@ -108,6 +111,27 @@ def write_properties_file(file_name) # rubocop:disable Metrics/PerceivedComplexi
...
+        sig { returns(T.nilable(Dependabot::DependencyFile)) }
+        def find_settings_file
+          settings_files = dependency_files.select do |f|
+            f.name == "settings.gradle" || f.name == "settings.gradle.kts"
+          end
+
+          # Prefer settings.gradle directly at job directory (File.dirname == directory)
+          # Otherwise choose shallowest by File.dirname depth (deterministic for composite builds)
+          settings_files.find { |f| File.dirname(f.path) == f.directory } ||
+            settings_files.min_by { |f| File.dirname(f.path).split("/").reject(&:empty?).length }
```
[the remark this answers is not in the document]
Comment (HaripriyaChintalapati) [reply gradle/lib/dependabot/gradle/file_updater/lockfile_updater.rb:133]: not required.

File: gradle/lib/dependabot/gradle/file_updater/lockfile_updater.rb [hunk 10]
```diff
@@ -108,6 +111,27 @@ def write_properties_file(file_name) # rubocop:disable Metrics/PerceivedComplexi
...
+        end
```
[the remark this answers is not in the document]
Comment (HaripriyaChintalapati) [reply gradle/lib/dependabot/gradle/file_updater/lockfile_updater.rb:134]: invalid

File: gradle/spec/dependabot/gradle/file_updater/lockfile_updater_spec.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (HaripriyaChintalapati) [reply gradle/spec/dependabot/gradle/file_updater/lockfile_updater_spec.rb:65]: in avid

File: gradle/lib/dependabot/gradle/file_updater/lockfile_updater.rb [hunk 11]
```diff
@@ -108,6 +111,27 @@ def write_properties_file(file_name) # rubocop:disable Metrics/PerceivedComplexi
...
         attr_reader :dependency_files
+
+        sig { params(build_file: Dependabot::DependencyFile).returns(String) }
+        def determine_root_dir(build_file)
+          settings_file = find_settings_file
+          return settings_file.directory if settings_file
+
+          # Fallback: use build file's job directory since it is always the project root
+          build_file.directory
+        end
```
[the remark this answers is not in the document]
Comment (HaripriyaChintalapati) [reply gradle/lib/dependabot/gradle/file_updater/lockfile_updater.rb:122]: invalid
